mne/channels/channels.py

```diff
@@ -1197,13 +1193,19 @@ def rename_channels(info, mapping):
         ch_names[c_ind] = new_name
 
     # check that all the channel names are unique
-    if len(ch_names) != len(np.unique(ch_names)):
+    if len(ch_names) != len(np.unique(ch_names)) and not allow_duplicates:
         raise ValueError('New channel names are not unique, renaming failed')
 
     # do the remapping in info
     info['bads'] = bads
+    rename = dict()
     for ch, ch_name in zip(info['chs'], ch_names):
+        rename[ch['ch_name']] = ch_name
         ch['ch_name'] = ch_name
+    _rename_comps(info.get('comps', []), rename)  # .get b/c fwd info omits it
+    for proj in info['projs']:
+        proj['data']['col_names'] = _rename_list(
+            proj['data']['col_names'].copy(), rename)
```

Comment:
this pattern is a bit misleading. The .copy() suggests it's inplace (which it is) but you sill affect the output.
If operation is inplace I would not return anything so it's not ambiguous